File: app/src/main/java/app/ds3wiki/character/CharacterAdapter.java
package app.ds3wiki.character;

import app.ds3wiki.R;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import static androidx.recyclerview.widget.RecyclerView.Adapter;
import static androidx.recyclerview.widget.RecyclerView.ViewHolder;

import androidx.annotation.NonNull;

final class CharacterAdapter extends Adapter<CharacterViewHolder> {
    private final CharacterRepository characterRepository = CharacterRepository.getInstance();

    @FunctionalInterface
    public interface OnClickListener {
        void onClick(final Character character);
    }

    private OnClickListener onClickListener;

    public void setOnClickListener(final OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    @NonNull
    @Override
    public CharacterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new CharacterViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.character_row_layout, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull CharacterViewHolder holder, int position) {
        final var root = holder.getRoot();
        final var context = root.getContext();

        root.setOnClickListener(v -> {
            onClickListener.onClick(characterRepository.getCharacter(position));
        });
    }

    @Override
    public int getItemCount() {
        return characterRepository.getRepositorySize();
    }
}

final class CharacterViewHolder extends ViewHolder {
    private final View root;

    public CharacterViewHolder(final View root) {
        super(root);
        this.root = root;
    }

    View getRoot() {
        return root;
    }
}

File: app/src/main/java/app/ds3wiki/characters/CharacterAdapter.kt
package app.ds3wiki.characters

import app.ds3wiki.R

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder

typealias CharacterViewHolder = CharacterAdapter.CharacterViewHolder

class CharacterAdapter : Adapter<CharacterViewHolder>() {
    lateinit var onCharacterSelect: (Character) -> Unit

    private val characterRepository = CharacterRepository.getInstance()

    inner class CharacterViewHolder(internal val view: View) : ViewHolder(view)

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CharacterViewHolder = CharacterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.character_row_layout, parent, false))

    override fun getItemCount(): Int = characterRepository.repositorySize

    override fun onBindViewHolder(
        holder: CharacterViewHolder,
        position: Int
    ) = holder.view.run {
        setOnClickListener {
            onCharacterSelect(characterRepository.getCharacter(position))
        }
    }
}